### ISSUE TITLE
[Discussion] Tag Usage & Tagging Strategy

### ISSUE THREAD

**Issue by BethanyG** [codebuddies/backend#73]
Please review this PR comment  as background:  https://github.com/codebuddies/backend/pull/69#discussion_r381446901

____

## The TL;DR

We are going to be tagging literally everything, and now is probably the time to have a discussion about our thoughts and architecture around it.

##  Some Background

*  We've decided to use a Django add-in library called [**`taggit`**](https://django-taggit.readthedocs.io/en/latest/index.html) to manage tagging for our backend API points.  We currently use the **default** configuration.

*  `Taggit`s default configuration uses Djangos [**contenttypes framework**](https://docs.djangoproject.com/en/3.0/ref/contrib/contenttypes/) and the Django ORMs [**generic relation**](https://docs.djangoproject.com/en/3.0/ref/contrib/contenttypes/#generic-relations) to model a **manytomany** relationship between **items to be tagged**  and **tags**.

*  We can change this behavior to be more direct and more explicit by using what is called a [**through model**](https://django-taggit.readthedocs.io/en/latest/custom_tagging.html), but this requires tracking tags for, say `Resources` in a `resources_tags` table, and tags for `Hangouts` in a `hangouts_tags` table. 

##  What this means

(_not an exhaustive list of points, but a start_)  

*  We can "tag" any model (resources, hangouts, people, discussions, notes) in a relatively straightforward matter by adding a "field" that points to `taggits` `TaggableManager` class.  See [this](https://django-taggit.readthedocs.io/en/latest/getting_started.html) for basic usage.  The `TaggableManager` takes care of looking up and saving the appropriate entries into the three associated "general" tables used for tagging.

*  We can use common `taggit` code to pull associated tags for an item, update those tags, or remove those tags, without having to write additional ORM code.

*  We can use common code to serialize and de-serialize tags for our API.  

*  Pulling tags across many items or endpoints could get **expensive**, due to the generic relations used.  See [**this warning from the docs**](https://django-taggit.readthedocs.io/en/latest/admin.html).

*  Someone cannot easily trace the relationship between the models being tagged and the tags/taggeditems/contenttyes from looking at the DB - there are no explicit links  from, for example, the resources_resource table to the taggit_tageditem table.

*   Custom serializer code had to be written to properly serialize the tags into the format we wanted them for for `Resources`.  We'll need to use the same code for any other endpoint that uses tags and that code may still have bugs in it.

## Some Questions

(_again, not exhaustive, but a start_)

*  Do we re-configure `taggit` to use explicit relations for each item being tagged?   Pros? Cons?
*  Do we want a explicit `Tags` endpoint to handle things like translating, filtering, pulling relations, etc? 

*  What makes sense for how we'll be using tags? 

*  Because of the queries involved, will we be setting ourselves up for performance issues?

*  Some additional related concerns in the conversation here:  https://github.com/codebuddies/backend/pull/69#discussion_r380910342

*  Some side debate about generic relations [**here**](https://lukeplant.me.uk/blog/posts/avoid-django-genericforeignkey/)

**Comment by lpatmo** [comment]
Thanks for these detailed notes and questions! :) Will let these questions sink in/give this some more thinking before I try to add to the discussion.

**Comment by BethanyG** [comment]
Additional thought about a common tags table:  context may or may not be harder to track.  As a "quick-and-dirty"  example, a "nighttime" tag may be relevant in the **Hangouts** context, but completely useless in the **Resources** context, and might have to be filtered out.  Or not.  Depending on how the tables/queries/endpoints are set up.

In any case, scenarios like the one above are what I'd love to suss out and discuss before we add in additional complexity for other endpoints.

**Comment by bengineerdavis** [comment]
Can we format tags as their own models? That way, we could create appropriate associations through foreign keys--let me know if I'm off on this.

**Comment by BethanyG** [comment]
@bengineerdavis  - TL;DR _**its complicated**_.  

Tags are already their own "model" -- they just aren't attached to what they are _tagging_ as a **Foreign Key** relation.  Not right now.

[**This Medium post**](https://medium.com/sthzg/a-short-exploration-of-django-taggit-bb869ea5051f) helped me think through things a bit.  As did this [**Django documentation on generic keys and contenttypes**](https://docs.djangoproject.com/en/2.2/ref/contrib/contenttypes/).

Finally, we went the route of customizing **`taggit`** (you can see the several failed PRs and the crazy stuff there) because we needed to make unicode render better in tags, and add a **`guid`** for each ....so [**this taggit documentation**](https://django-taggit.readthedocs.io/en/latest/custom_tagging.html#) goes through a bit of that.

And it was ... **_involved_**, at least it was for me.

**Comment by BethanyG** [comment]
Closing this in favor of a discussion.
